modules/ROOT/pages/ai-gemini.adoc

```diff
@@ -0,0 +1,224 @@
+// Instead, an alternate method for retrieving these credentials should be used.
+const ACCESS_TOKEN = '<INSERT_ACCESS_TOKEN_HERE>'; // used in the Authorization Header
+const PROJECT_ID = '<INSERT_PROJECT_ID_HERE>'; // retrieve from the Vertex AI Google Console
+const REGION = '<INSERT_REGION_HERE>'; // e.g. us-central1
+const API_URL = `https://${REGION}-aiplatform.googleapis.com/v1/projects/${PROJECT_ID}/locations/${REGION}/publishers/google/models/gemini-1.0-pro:streamGenerateContent?alt=sse`;
+
+const ai_request = (request, respondWith) => {
+  respondWith.stream((signal, streamMessage) => {
+	// Adds each previous query and response as individual messages
+	const conversation = request.thread.flatMap((event) => {
+	  if (event.response) {
+		return [
+		  { role: 'user', parts: [{ text: event.request.query }] },
+		  { role: 'model', parts: [{ text: event.response.data }] }
+		];
+	  } else {
+		return [];
+	  }
+	});
+
+	// System messages provided by the plugin to format the output as HTML content.
+	const pluginSystemMessages = request.system.map((text) => ({
+	  text
+	}));
+
+	const systemInstruction = {
+	  parts: [
+		...pluginSystemMessages,
+		// Additional system messages to control the output of the AI
+		{ text: 'Do not include html``` at the start or ``` at the end of the response.' },
+		{ text: 'No boilerplate or explanation, just give the HTML response.' }
+	  ]
+	};
+
+	// Forms the new query sent to the API
+	const text = request.context.length === 0 || conversation.length > 0
+	  ? request.query
+	  : `Question: ${request.query} Context: """${request.context}"""`;
+
+	const contents = [
+	  ...conversation,
+	  { role: 'user', parts: [
+		{ text }
+	  ]}
+	];
+
+	const generationConfig = {
+	  temperature: 0.9,
+	  maxOutputTokens: 800,
+	};
+
+	const requestBody = {
+	  contents,
+	  generationConfig,
+	  systemInstruction,
+	};
+
+	const geminiOptions = {
+	  method: 'POST',
+	  headers: {
+		'Content-Type': 'application/json',
+		'Authorization': `Bearer ${ACCESS_TOKEN}`
+	  },
+	  body: JSON.stringify(requestBody)
+	};
+
+	const onopen = async (response) => {
+	  if (response) {
+		const contentType = response.headers.get('content-type');
+		if (response.ok && contentType?.includes('text/event-stream')) {
+		  return;
+		} else if (contentType?.includes('application/json')) {
+		  const data = await response.json();
+		  if (data.error) {
+			throw new Error(`${data.error.type}: ${data.error.message}`);
+		  }
+		}
+	  } else {
+		throw new Error('Failed to communicate with the Gemini API');
+	  }
+	};
+
+	// This function passes each new message into the plugin via the `streamMessage` callback.
+	const onmessage = (ev) => {
+	  const data = ev.data;
+	  if (data !== '[DONE]') {
+		const parsedData = JSON.parse(data);
+		const message = parsedData?.candidates[0]?.content?.parts[0]?.text?.replace(/^```html\n/, "").replace(/\n```$/, "");
```

Comment:
We don't need to include the markdown replacement code.

Reply:
I found that without the replace statements, the response contains the markdown formatting:
<img width="1222" alt="image" src="https://github.com/tinymce/tinymce-docs/assets/40410898/03ce4def-d45e-4663-9bff-e1d12e512777">

Are you okay with this?

Reply:
That's a string response, I don't think that workaround will work in a streaming response

Reply:
Based on my testing, it seems to work with streaming too (see console).

<img width="1649" alt="image" src="https://github.com/tinymce/tinymce-docs/assets/40410898/1d9c50be-e6ae-438a-836e-c79a9abfdacd">

Reply:
Looks good then. It seems Gemini consistently chunks that text in a single message, where OpenAI would often separate it into a few messages.

Reply:
Yes, OpenAI's responses are much more frequent and fragmented. Gemini returns much larger chunks, so the markdown formatting almost always ends up in the first and last chunks.